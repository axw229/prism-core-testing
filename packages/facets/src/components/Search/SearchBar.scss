.SearchBar {
  $input-bg: var(getThemeColor('white'));
  $input-padding: 0.5em;
  $clear-size: 1em;
  font-family: var(getVarName('typography.titleFontFamily'));

  &__search-form {
    width: 100%;
  }

  &__label {
    cursor: default;
    display: inline-block;
    font-size: 0.9em;
    font-weight: bold;
    margin-bottom: 0.25em;
  }

  &__icon {
    margin: 0 $ui-spacing 0 0;
  }

  &__cancel-button {
    margin: 0 0 0 $ui-spacing;
  }

  &__inner {
    display: flex;
    flex-direction: row;
    align-items: center;
    padding: $ui-spacing;
  }

  &__wrapper {
    background: $input-bg;
    position: relative;
    margin: 0;
    flex: 1;
    height: $ui-height-min--mobile;
    border-radius: var(getVarName('typography.textInputBorderRadius'));

    @include respond-above(xs) {
      height: $ui-height-min;
    }

    &--without-outline {
      border: 1px solid $color-light-grey;
    }

    &--with-outline {
      border: 1px solid var(getThemeColor('primary'));
      box-shadow: inset 0 0 0 1px var(getThemeColor('primary'));
    }
  }

  &__input {
    outline: none;
    border: none;
    width: 100%;
    height: 100%;
    position: absolute;
    line-height: $ui-height-min--mobile;
    text-indent: $input-padding;
    padding: 0 $clear-size + (2 * $input-padding) 0 0;

    @include respond-above(xs) {
      line-height: $ui-height-min;
    }
  }

  .SearchBar__input::placeholder {
    color: var(getThemeColor('primaryPlaceholder'));
  }

  @media (max-width: 420px) {
    .SearchBar__input::placeholder {
      font-size: 1em;
    }
  }

  @media (max-width: 380px) {
    .SearchBar__input::placeholder {
      font-size: 0.8em;
    }
  }

  &__clean {
    position: absolute;
    right: ($clear-size / 2) + $input-padding;
    top: 50%;
    transform: translateY(-50%) translateX(50%);
    font-size: $clear-size;
    width: $clear-size * 1.5;
    height: $clear-size * 1.5;
    border-radius: 50%;

    &:hover,
    &:focus {
      box-shadow: inset 999em 0 0 0 rgba(white, 0.9);
      background: var(getThemeColor('primary'));
      color: var(getThemeColor('dark'));
    }
  }
}

.SearchBarLight {
  &__wrapper {
    border: none;
    color: #404040;
  }

  &__input::placeholder {
    color: #404040;
    font-size: inherit !important;
  }

  &__back {
    font-size: 1rem;
    margin-right: $ui-spacing;
  }

  &__icon {
    font-size: 12px;
    margin: 0 0 0 $ui-spacing;
    order: 3;
  }
}

.SearchBarMinimal {
  .button-bar__options__option__btn {
    background: none !important;
    color: black !important;
    text-transform: none;
  }

  .button-bar__options__option__btn,
  .SearchBar__input {
    font-size: 17px;
    font-weight: 600;
    line-height: 24px;
  }

  .button-bar__option-container {
    border: none !important;
  }
}
