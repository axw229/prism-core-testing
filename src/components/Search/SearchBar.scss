.SearchBar {
  $input-bg: white;
  $input-padding: .5em;
  $clear-size: 1em;

  &__search-form {
    width: 100%;
  }

  &__label {
    cursor: default;
    display: inline-block;
    font-size: .9em;
    font-weight: bold;
    margin-bottom: .25em;
  }

  &__icon {
    margin: 0 $ui-spacing 0 0;
  }

  &__cancel-button {
    margin: 0 0 0 $ui-spacing;
  }

  &__inner {
    display: flex;
    flex-direction: row;
    align-items: center;
    padding-bottom: $ui-spacing;
  }

  &__wrapper {
    background: $input-bg;
    position: relative;
    margin: 0;
    flex: 1;
    height: $ui-height-min--mobile;
    border-radius: $ui-radius;

    @include respond-above(xs) {
      height: $ui-height-min;
    }

    &--without-outline {
      border: 1px solid $color-light-grey;
    }

    &--with-outline {
      border: 1px solid var(getThemeColor('primary'));
      box-shadow: inset 0 0 0 1px var(getThemeColor('primary'));
    }
  }

  &__input {
    outline: none;
    border: none;
    width: 100%;
    height: 100%;
    position: absolute;
    line-height: $ui-height-min--mobile;
    text-indent: $input-padding;
    padding: 0 $clear-size + (2 * $input-padding) 0 0;

    @include respond-above(xs) {
      line-height: $ui-height-min;
    }
  }

  &__clean {
    position: absolute;
    right: ($clear-size / 2) + $input-padding;
    top: 50%;
    transform: translateY(-50%) translateX(50%);
    font-size: $clear-size;
    width: $clear-size * 1.5;
    height: $clear-size * 1.5;
    border-radius: 50%;

    &:hover, &:focus {
      box-shadow: inset 999em 0 0 0 rgba(white, .9);
      background: var(getThemeColor('primary'));
      color: var(getThemeColor('dark'));
    }
  }
}
