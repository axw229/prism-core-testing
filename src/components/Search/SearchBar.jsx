/* eslint-disable jsx-a11y/label-has-for */
// @flow
import React from 'react'
import { useHistory, useParams } from 'react-router-dom'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import 'src/providers/fontawesome/fontawesome'
import ButtonBar from '../GeneralButtons/ButtonBar/ButtonBar'
import { FormattedMessage } from 'react-intl'
import uniqueId from 'lodash/uniqueId'
import debounce from 'lodash/debounce'
import './SearchBar.scss'
import 'src/scss/convenience/visually-hidden.scss'
import { MIN_SEARCH_LENGTH } from '../../store/actions/loadSearchResults'

type Props = {
  showCancelButton: boolean,
  showLabel?: boolean,
  showIcon?: boolean,
  label: string,
  placeholder?: string
}

export default (props: Props) => {
  const {
    label,
    placeholder,
    showCancelButton = true,
    showIcon = true,
    showLabel = true
  } = props
  const { query = '' } = useParams()
  const history = useHistory()
  const [value, setValue] = React.useState(query)
  const [id] = React.useState(uniqueId('SearchBarInput'))
  const updateUrl = React.useMemo(() => debounce((value, abort = false) => {
    // this is for "cancelling" the debounced method if we unmount before execution
    if (abort) {
      return
    }

    if (typeof value === 'string') {
      const l = value.length
      if (l >= MIN_SEARCH_LENGTH) {
        // ensure we aren't pushing anything dangerous into the URL
        const v = encodeURIComponent(value)

        // if this is a new search query...
        if (value !== query) {
          // ... push it to the URL
          history.push(v)
        } else {
          // ... otherwise just replace the current URL with it
          // this enables searching on initial load w/ a search param
          history.replace(v)
        }
      } else if (l === 0 && value !== query) {
        // this allows us to escape via routing if the user clears their search
        // don't allow if query is already blank -- doing it then auto-pushes a blank state on initial load
        history.push('./')
      }
    }
  }, 250), [query])

  React.useEffect(() => { // mutate url when input hasn't changed in 250ms
    updateUrl(value)

    return () => {
      updateUrl(null, true)
    }
  }, [value])

  return (
    <div className='SearchBar'>
      <form onSubmit={e => e.preventDefault()} className='SearchBar__search-form'>
        <label className={`SearchBar__label ${!showLabel ? 'visually-hidden' : ''}`} htmlFor={id}>{label}</label>
        <div className='SearchBar__inner'>
          {showIcon && (<label htmlFor={id} className='SearchBar__icon'>
            <FontAwesomeIcon icon={['fal', 'search']} size='lg' />
          </label>)}
          <div className={`SearchBar__wrapper SearchBar__wrapper--with${query ? '-outline' : 'out-outline'}`}>
            <input id={id} value={value} className='SearchBar__input' onChange={e => setValue(e.target.value)} placeholder={placeholder} />
            {value.length > 0 &&
              <button type='button' className='SearchBar__clean' onClick={() => setValue('')}>
                <FontAwesomeIcon icon={['fas', 'times']} />
              </button>}
          </div>
          {showCancelButton && <div className='SearchBar__cancel-button'>
            <ButtonBar.Bar>
              <ButtonBar.Button to='../'>
                <FormattedMessage id='CANCEL' />
              </ButtonBar.Button>
            </ButtonBar.Bar>
          </div>}
        </div>
      </form>
    </div>
  )
}
