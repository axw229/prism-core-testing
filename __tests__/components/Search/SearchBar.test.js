/* eslint-env jest */
import React from 'react'
import { shallow } from 'enzyme'
import SearchBar from 'src/components/Search/SearchBar'

const mockFnOnSearchInput = jest.fn()
const mockFnOnClearSearch = jest.fn()
const mockHandleSearchInput = jest.fn()
const mockPersist = jest.fn()

const getSearch = (props) => {
  const defaultProps = {
    onSearchInput: mockFnOnSearchInput,
    onClearSearch: mockFnOnClearSearch
  }
  const newProps = Object.assign({}, defaultProps, props)
  return shallow(<SearchBar {...newProps} />)
}

describe('SearchBar', () => {
  let search
  beforeEach(() => {
    if (!search) {
      search = getSearch()
    }
  })

  it('should match snapshot', () => {
    expect(search).toMatchSnapshot()
  })

  it('should render input', () => {
    expect(search.find('input').exists()).toBe(true)
  })
})

describe('SearchBar with events', () => {
  let search
  beforeEach(() => {
    if (!search) {
      search = getSearch()
    }
  })

  xit('should call handleInput on input change', () => {
    search.instance().handleInput = mockHandleSearchInput
    search.instance().forceUpdate()
    search.find('input').simulate('input', { target: { value: 'xyz' }, persist: mockPersist })
    expect(mockHandleSearchInput).toHaveBeenCalledWith('xyz')
  })
})
