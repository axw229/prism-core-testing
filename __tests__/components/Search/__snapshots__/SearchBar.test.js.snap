// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar should match snapshot 1`] = `
<ForwardRef
  onClearSearch={[MockFunction]}
  onSearchInput={[MockFunction]}
>
  <SearchBar
    forwardedRef={null}
    onClearSearch={[MockFunction]}
    onSearchInput={[MockFunction]}
  >
    <div
      className="SearchBar"
    >
      <FontAwesomeIcon
        border={false}
        className="search-icon"
        fixedWidth={false}
        flip={null}
        icon={
          Object {
            "icon": Array [
              512,
              512,
              Array [],
              "f002",
              "M508.5 481.6l-129-129c-2.3-2.3-5.3-3.5-8.5-3.5h-10.3C395 312 416 262.5 416 208 416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c54.5 0 104-21 141.1-55.2V371c0 3.2 1.3 6.2 3.5 8.5l129 129c4.7 4.7 12.3 4.7 17 0l9.9-9.9c4.7-4.7 4.7-12.3 0-17zM208 384c-97.3 0-176-78.7-176-176S110.7 32 208 32s176 78.7 176 176-78.7 176-176 176z",
            ],
            "iconName": "search",
            "prefix": "fal",
          }
        }
        inverse={false}
        listItem={false}
        mask={null}
        pull={null}
        pulse={false}
        rotation={null}
        size="lg"
        spin={false}
        symbol={false}
        title=""
        transform={null}
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-search fa-w-16 fa-lg search-icon"
          data-icon="search"
          data-prefix="fal"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M508.5 481.6l-129-129c-2.3-2.3-5.3-3.5-8.5-3.5h-10.3C395 312 416 262.5 416 208 416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c54.5 0 104-21 141.1-55.2V371c0 3.2 1.3 6.2 3.5 8.5l129 129c4.7 4.7 12.3 4.7 17 0l9.9-9.9c4.7-4.7 4.7-12.3 0-17zM208 384c-97.3 0-176-78.7-176-176S110.7 32 208 32s176 78.7 176 176-78.7 176-176 176z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </FontAwesomeIcon>
      <div
        className="SearchBar__wrapper SearchBar__wrapper--without-outline"
      >
        <input
          className="SearchBar__input"
          onChange={[Function]}
        />
      </div>
    </div>
  </SearchBar>
</ForwardRef>
`;
